kimchi/src/circuits/polynomials/not.rs

```diff
@@ -0,0 +1,235 @@
+//! This module includes the definition of the NOT gadget and the witness code generation,
+//! for both the implementation running with `Xor16` gates and the one with `Generic` gates.
+//! Note that this module does not include a `Not` gate type.
+use crate::circuits::{
+    gate::{CircuitGate, Connect, GateType},
+    polynomial::COLUMNS,
+    wires::Wire,
+};
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use o1_utils::{BigUintHelpers, BitwiseOps, FieldHelpers};
+use std::{array, cmp::max};
+
+use super::{
+    generic::GenericGateSpec,
+    xor::{init_xor, num_xors},
+};
+
+//~ We implement the NOT gadget making use of the XOR gadget and the Generic gate, in two different ways. A new gate type is not needed.
+//~
+//~ The first version of the NOT gadget reuses `Xor16` by making the following observation:
+//~ $$\textit{the bitwise NOT operation is equivalent to the bitwise XOR operation with the all one words of a certain length}$$
+//~ Then, if we take the XOR gadget with a second input to be the all one word of the same length, that gives us the NOT gadget.
+//~ The correct length can be imposed by having a public input containing the `2^bits - 1` value and wiring it to the second input of the XOR gate.
+//~ This approach needs as many rows as a XOR would need, for a single negation, but it comes with the advantage of making sure the input is of a certain length.
+//~
+//~ The other approach can be more efficient if we already know the length of the inputs. For example, the input may be the input of a range check gate,
+//~ or the output of a previous XOR gadget (which will be the case in our Keccak usecase).
+//~ In this case, we simply perform the negation as a subtraction of the input word from the all one word (which again can be copied from a public input).
+//~ This comes with the advantage of holding up to 2 word negations per row (an eight-times improvement over the XOR approach), but it requires the user to know the length of the input.
+
+impl<F: PrimeField> CircuitGate<F> {
+    /// Creates a negation gadget with `nots` NOT components of some length previously constrained using a generic gate
+    /// (checking that a cell stores `2^bits-1` value). Assumes that the inputs are known to have at most `bits` length.
+    /// Starts the gates in the `new_row` position.
+    /// Includes:
+    /// - ceil(nots/2) Double Generic gates to perform the `( 2^(bits) - 1 ) - input` operation for every two inputs in each row
+    /// BEWARE:
+    /// - If the input is not known to have at most `bits` length, this should be constrained somewhere else.
+    /// - Otherwise, use the Xor builder instead (but this one requires about 8 times more rows).
+    /// INTEGRATION:
+    /// - Needs a leading public input generic gate in `pub_row` to constrain the left input of each generic gate for negation to be `2^bits-1`.
+    pub fn extend_not_gnrc_gadget(
```

Comment:
wow that is a long name... I guess you don't like the former because it may seem like you're negating the "un-checked" thus reading like "checked"?

Reply:
I may use `neg` instead of not then, to avoid confusion, and use the unchecked/checked approach to abstract away the implementation details of each of them

Reply:
I'm just trying to find a name that makes it clear that one gate requires an external check of the length in order to be sound.   This is sorta analogous to how some rust functions has an explicit `_unsafe` in their name... so something like that just to make it clear that using this gate requires you to do something else to be **safe**